protocol/decoder/decoding_error.go

```diff
@@ -0,0 +1,22 @@
+package decoder
+
+// decodingError is an error that occurs in the process of decoding
+// For eg. no bytes remaining, invalid bytes in varint, etc.
+// For this, the start and end offset is the same value because we only point to the
+// byte from where the error was encountered while trying to decode a value
+type decodingError struct {
```

Comment:
I'd stick to the single interface and just have two constructors: 

```
NewDecoderErrorForOffset()
NewDecoderErrorForRange()
```

We don't really need two different types here

Reply:
Also avoid the term "decodingError" - stick to `decoderError` like before (the smaller-case version of the interface)